samples/mesh: Fix dev_uuid initialization from identity address

jhedberg opened the pull request:
With the recent changes to identity address handling the local
identity address is only guaranteed to be available once
settings_load() has been called. Move the initialization of dev_uuid
to the appropriate place.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
samples/mesh: Fix dev_uuid initialization from identity address

With the recent changes to identity address handling the local
identity address is only guaranteed to be available once
settings_load() has been called. Move the initialization of dev_uuid
to the appropriate place.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
@vikrant8051 this should fix the issue with the UUID that you reported. Could you please verify that it does indeed fix it?

vikrant8051 commented:
@jhedberg Ok.

vikrant8051 commented:
@jhedberg Still getting UUID as 00000000000000000000000000000000

jhedberg commented:
> @jhedberg Still getting UUID as 00000000000000000000000000000000

I the Mesh stack debug logs, or in the actual unprovisioned device beacons? Since the code is now calling bt_mesh_init() before setting the UUID the Mesh stack logs are expected to print all zeroes in this case (which you can just ignore).

vikrant8051 commented:
Ofcourse while DEVICE is in unprovisioned state

Commit:
Bluetooth: Mesh: Move Device UUID log to bt_mesh_prov_enable()

In some cases the application might only initialize its UUID after
calling bt_mesh_init(), e.g. in the case of deriving the UUID from the
identity address. To avoid confusing logs, only print the UUID when
actually enabling one of the provisioing bearers.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
> Ofcourse while DEVICE is in unprovisioned state

What do you mean?

Anyway, I just pushed a second patch to move the confusing log (which I suspect you're seeing) to a better place.

jhedberg commented:
> Ofcourse while DEVICE is in unprovisioned state

@vikrant8051 I wasn't asking *when* you get the all-zeroes UUID, rather *where* you are seeing it. I suspect it was only in the log but over the air the correct value was used. As I said, my second patch should also fix the logging (hopefully).

vikrant8051 commented:
@jhedberg I do same changes as per your this commit & execute 

[meshctl]# discover-unprovisioned on
SetDiscoveryFilter success
Discovery started
Adapter property changed 
[CHG] Controller 28:F0:76:2F:42:BB Discovering: yes
		Mesh Provisioning Service (00001827-0000-1000-8000-00805f9b34fb)
			Device UUID: **00000000000000000000000000000000**
			OOB: 0000

Previously here we could see starting bytes as MAC address.

vikrant8051 commented:
@jhedberg 
After setting static u8_t dev_uuid[16] = {0xdd, 0xdd}; now I got 
UUID as 			Device UUID: dddd0000000000000000000000000000

jhedberg commented:
@vikrant8051 ok, there must be something I've missed then. I'll try to reproduce myself.

jhedberg commented:
@vikrant8051 I can reproduce. The issue exists for PB-GATT but not for PB-ADV. The bt_mesh_proxy_init() function is constructing the advertising data and doesn't re-check the UUID later when actually using it. I'll try to come up with a fix soonish.

Commit:
Bluetooth: Mesh: Initialize PB-GATT advertising data at the right time

Some things, such as UUID, URI or even the local name may be different
when bt_mesh_prov_enable() is called compared to when bt_mesh_init()
was called. Create the advertising data on-demand each time when
enabling PB-GATT.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
@vikrant8051 please try again with this PR. There's now a third patch that should fix this for PB-GATT.

vikrant8051 commented:
Ok

vikrant8051 commented:
@jhedberg Nope !! Still getting  Device UUID: dddd0000000000000000000000000000 after setting 
static u8_t dev_uuid[16] = {0xdd, 0xdd}; or 00000000000000000000000000000000

git log has shown me following commits in my local repo ...
commit 22f57fb121516ddc2f8b600fcfbe39b3140c02ee (HEAD -> fix_bugs10)
commit 8e4982808637045f7bca4e6ad3d7f7d1e07c9c50
commit 42f1972077cc913a30662707c4b707e94c0f8008

jhedberg commented:
@vikrant8051 I'm not following. If you're doing that initialization then you *are* supposed to get dddd. Remove the static initialization and keep the identity address based code as is. FWIW, I tried this myself with an nRF52840 DK and I was able to see the correct UUID in the nRFConnect app.

jhedberg commented:
@vikrant8051 I cannot reproduce what you are saying. With the latest patches in this PR, on an nRF52840 DK, when I scan from my phone (nRF Connect) it shows that the UUID has been correctly initialized from the identity address.

vikrant8051 commented:
@jhedberg 
Do we have to enable some configuration option in prj.conf to use MAC address intead of value defined to dev_uuid ?

jhedberg commented:
> Do we have to enable some configuration option in prj.conf to embed MAC address in UUID instead of using value defined for dev_uuid ?

@vikrant8051 No, it's all app-side. You just need to have the memcpy as-is in the code. Did you really try this PR without any other changes to it?

vikrant8051 commented:
@jhedberg 
Issue has solved ... it was because of my /etc/bash.bashrc which was executing as usual script from /home/vikrant/zephyr/zephyr-env.sh

Now I changed it to latest location where I added this PR.

vikrant8051 commented:
@jhedberg 
Please take care of things if unnecessary changes has happened due to wrong feedback from my side.

jhedberg commented:
>  @jhedberg Please take care of things if unnecessary changes has happened due to wrong feedback from my side.

I don't think there's anything unnecessary in the PR, since I was able to reproduce the issue myself and verify that the patches fix the behavior. I think all three patches are good. Now I'd just need an approved review from someone, @carlescufi @Vudentz @sjanc ? :)

vikrant8051 commented:
@jhedberg Understanding Mesh stack is beyond my scope. I never tried to judge or do any analysis what is going on at background. It was just my concern since it was my mistake.

cvinayak reviewed (approved):
LGTM
